### PR TITLE
Code normalization script

### PR DESCRIPTION
**Pull request by bobbicodes**
After [this livestream](https://www.youtube.com/watch?v=dF-bsf9iweE), I'd say we are about halfway done with the representer, and I believe the most difficult part is over because I figured out a workable approach to editing the source code tree. 

Here's what we came up with:

``` clojure
(defn arglists [var-def]
  (let [lists (map edn/read-string (:arglist-strs var-def))]
    (into [] (for [list lists]
               {:var  (:name var-def)
                :args list}))))
```
This takes a single var definition from the analyzer, which contains the arglists for each arity of the function being defined. It then outputs a sequence of maps, each containing an arglist as well as the name of the var. This way we can use a reducing function over the collection:

``` clojure
(defn replace-arglist [z arglist]
    (-> z
        (z/find-value z/next (:var arglist))
        (z/find-next-depth-first #(= (:args arglist) (z/sexpr %)))
        (z/replace (mapv #(symbol (get placeholders (str %)))
                         (:args arglist)))))
```
The trick here is the predicate function `#(= (:args arglist) (z/sexpr %))` which is passed to `z/find-next-depth-first` which navigates the zipper to the node which contains the arglists being replaced, which we know is located in the correct function because we navigated to that in the previous line. It then builds a new vector containing the right placeholder names by looking them up.

We then pass that function to `reduce` along with the complete collection of arglist maps:

``` clojure
(-> (reduce replace-arglist impl 
            (mapcat arglists (:var-definitions analysis)))
    z/root-string
    println)
```

Output:

``` clojure
(defn with-name
  ([PLACEHOLDER-1]
   (with-name name "white"))
  ([PLACEHOLDER-1 PLACEHOLDER-2]
   (let [my-fn (fn inline-fn [name2 color2]
                 (str name2 " " color2))]
     (str "My name is " name
          " and my color is " color
          " and my-fn is " (my-fn "joe" "blue")))))
```

This might seem like an awful lot of fuss just to replace those 3 names, but that's because we had to find a reliable way to replace _only_ the right ones!

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,5 +1,29 @@
...
+# Install SSL ca certificates
+RUN apk update && apk add ca-certificates
+
+# Create appuser
+RUN adduser -D -g '' appuser
+
+# get the source code
+WORKDIR /clojure-representer
+COPY clojure_representer.clj .
+COPY bin/run.sh bin/run.sh
```
Comment (ErikSchierboom) [inline Dockerfile]: You could consider swapping these two lines to increase the likelihood of Docker caching being used, as I assume `bin/run.sh` is more stable than the representer code itself.

File: Dockerfile [hunk 2]
```diff
@@ -1,5 +1,29 @@
...
+
+# Get static analyzer image as build stage
+FROM cljkondo/clj-kondo:2021.01.20-alpine as clj-kondo
+
+# Build a minimal and secured container
+FROM babashka/babashka:0.10.163-alpine
+COPY --from=builder /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/
+COPY --from=builder /etc/passwd /etc/passwd
+COPY --from=builder /clojure-representer/clojure_representer.clj /opt/representer/clojure_representer.clj
+COPY --from=builder /clojure-representer/bin/run.sh /opt/representer/bin/run.sh
```
Comment (ErikSchierboom) [inline Dockerfile]: I think you could also do this.

```suggestion
WORKDIR /opt/test-runner
COPY --from=build /clojure-representer/ .
```

Which means that you could move the `COPY` command after the existing `WORKDIR /opt/representer` line.

File: Dockerfile (hunk 1 above)
Comment (ErikSchierboom) [reply Dockerfile]: Secondly, as these commands don't _need_ to be in the builder, you could also just directly add them to the non-builder stage I think.

File: Dockerfile [hunk 3]
```diff
@@ -1,5 +1,29 @@
...
+COPY clojure_representer.clj .
+COPY bin/run.sh bin/run.sh
+
+# Get static analyzer image as build stage
+FROM cljkondo/clj-kondo:2021.01.20-alpine as clj-kondo
+
+# Build a minimal and secured container
+FROM babashka/babashka:0.10.163-alpine
+COPY --from=builder /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/
+COPY --from=builder /etc/passwd /etc/passwd
```
Comment (ErikSchierboom) [inline Dockerfile:20]: I assume you're using the builder to not have the overhead of `apk update` within this container? If so, I think there are also options to not have apk update cache or to empty its cache. That way you wouldn't need the builder image. But it's fine to keep it like this of course, just curious.

File: bin/run-in-docker.sh [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+# Usage:
+# ./bin/run-in-docker.sh two-fer ~/folder/to/solution/input ~/folder/to/representer/output
+
+# If any required arguments is missing, print the usage and exit
+if [ -z "$1" ] || [ -z "$2" ] || [ -z "$3" ]; then
+    echo "usage: ./bin/run-in-docker.sh exercise-slug path/to/solution/folder/ path/to/output/directory/"
+    exit 1
+fi
+
```
Comment (ErikSchierboom) [inline bin/run-in-docker.sh]: Might I suggested adding a `set -e`:

```suggestion
set -e

```

We have that included in the latest version of the generic test runner, but many of the existing tooling repos don't yet have it. This has caused issues in the past, as you could have a failing `docker build` command but still have the `docker run` succeed as that would then download the latest version from Docker hub, not using the (broken) local image.

See https://github.com/exercism/generic-test-runner/blob/main/bin/run-in-docker.sh#L19-L20

**Review by ErikSchierboom — APPROVED** [review]
Great job! Is there code that verifies that the representer works? I might just be missing it whilst reading.

File: Dockerfile (hunk 3 above)
Comment (bobbicodes) [reply Dockerfile:20]: I'm actually curious too, my assumption was that it had to do with security. I just saw that's what everyone else was doing 🐒

File: bin/run-in-docker.sh (hunk 4 above)
Comment (bobbicodes) [reply bin/run-in-docker.sh]: Oh I like that, done.

File: Dockerfile (hunk 3 above)
Comment (ErikSchierboom) [reply Dockerfile:20]: It could be 🤷

**Review by ErikSchierboom — APPROVED** [review]

**Comment by bobbicodes** [comment]
Superseded by #26

**Comment by bobbicodes** [comment]
Reopening this because it seems it may be an approach worth developing further.

The key difference is that it uses static analysis which avoids the complexity involved with executing code. This is not due to a security concern as one might expect, as I assume that our containerization takes care of any dangers of running untrusted code, it just turns out that Clojure's implementation details (being hosted on the JVM) make the process more involved than I had expected. In particular, if a solution happens to use an external library, this means it needs to be dynamically resolved at runtime and added to the classpath.

To get a hint of what I mean, check out [The Classpath is a Lie](https://lambdaisland.com/blog/2021-08-25-classpath-is-a-lie) by Arne Brasseur. It's enough to get me running away like a scared little kitty cat with my tail between my legs.

I have taken to labeling this condition "The Classpath Blues", and it will be released on my next album.

**Review by ErikSchierboom — APPROVED** [review]

**Comment by bobbicodes** [comment]
I realized that we need access to the exercise test suite in order to determine which function names should be normalized, and which ones should not because they are tested and thus an essential part of the solution.

For example, solutions often include intermediate functions, like this:
``` clj
(defn to-string [x]
  (str x))
```

and another solution might include effectively the same function with a different name:
``` clj
(defn ->string [x]
  (str x))
```

Actually this happens quite often - looking at `armstrong-numbers`, since Clojure has no built-in exponent function (with the exception of Java's `Math.pow`, which is seldom used because it has strange behavior with doubles which needs to be specially handled), most students write their own, which I see named `pow`, `power`, `exp`, `expt`, `raise-digits-to-length`, `**`, `pow-pow`... you get the idea, but the implementation is nearly always `(reduce * (repeat n x))`.

 so it would be good to normalize these, but we need to _not_ do so for functions that are _not_ intermediate.

**Comment by bobbicodes** [comment]
This repo now contains an _obscene_ number of test files and this should not be necessary. They can be drastically reduced if I simply remove the solutions that are exact duplicates (but I'll leave 2 of them so we know the representer correctly handles duplicates)

EDIT: Actually for the purpose of testing, it's probably not necessary for the solutions to even be in separate files with the full project directory structure. I might as well just create a single file with the set of unique solutions in it.

Rationale: it's not necessary to be testing the file i/o machinery 1000 times, we could just do that once

But wait... if I remove the solutions from their original files, that will break the analysis info which relies on position data. Though I'm actually leaning towards not using that anyway so I don't know if I care about that. And I suppose the solution string still contains all the same formatting so that's probably not an issue.

This is kind of crazy... the original directory of twofers is 18.8 MB, and the `unique-twofers` file I just made is 57KB

**Comment by bobbicodes** [comment]
Closing this even though it was loads of fun! For awhile I really thought this would be the way to go, but [eastwood](https://github.com/jonase/eastwood) got me convinced that if you can afford to wait a few seconds, evaluation/macroexpansion is 👑

**Pull request closed without merging**